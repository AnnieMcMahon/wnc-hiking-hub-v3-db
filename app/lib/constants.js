//TRAIL SEARCH CRITERIA
export const ANY_AREA = "Anywhere in WNC";
export const ANY_DIFFICULTY = "Any difficulty";
export const ANY_LENGTH = "Any length";

export const SHORT = "Shorter than 3 miles";
export const MEDIUM = "From 3 to 6 miles";
export const LONG = "Longer than 6 miles";

export const AREAS = [
  "Bent Creek Experimental Forest",
  "Blue Ridge Parkway",
  "DuPont State Recreational Forest",
  "Gorges State Park",
  "North Carolina Arboretum",
  "Pisgah National Forest",
  "Other (not listed)"
];

export const DIFFICULTIES = [
  "easy",
  "moderate",
  "hard"
]

export const LENGTHS = [
  SHORT,
  MEDIUM,
  LONG
];

export const ROUTE_TYPES = [
  "loop",
  "out-and-back",
  "point-to-point"
];

//BIO INFO
export const DEFAULT_USER_NAME = "New User";
export const DEFAULT_AVATAR = "/newUser.png";
export const DEFAULT_BIO = "Tell us about yourself";

//DEFAULT USER WHEN NO ONE IS SIGNED IN
export const DEFAULT_USER = {
  id: 1,
  email: "demo@gmail.com",
  user_name: "Demo User",
  avatar: "/newUser.png",
  bio: "Bio text goes here ... Lorem ipsum dolor sit amet consectetur adipisicing elit. Rem reprehenderit assumenda quibusdam iste eos exercitationem molestias nulla fuga repellat incidunt ullam cupiditate necessitatibus quidem debitis dolorum nesciunt labore quos, laboriosam tempora qui tempore quo velit! Libero quod eligendi at eum.",
};

export const BLANK_USER = {
  id: 0,
  email: "",
  user_name: "New User",
  avatar: "/newUser.png",
  bio: "Tell us about yourself"
};

export const BLANK_BIO = {
  user_name: "",
  bio: ""
};

export const BLANK_HIKE = {
  id: 0,
  title: "",
  date: "",
  time: "",
  location: "",
  comments: "",
  creator_id: 0,
  status: ""
};

export const BLANK_HIKE_INFO = {
  id: 0,
  title: "",
  date: "",
  time: "",
  location: "",
  comments: "",
  status: "",
  creator_id: 0,
  creator: "",
  buttonMessage: ""
};

export const BLANK_TRAIL = {
  id: 0,
  trail_name: "",
  area_name: "",
  difficulty_rating: "",
  length: 0,
  elevation_gain: 0,
  route_type: "",
  trail_link: ""
};

export const MOCK_TRAIL_INFO = {
  id: 5,
  trail_name: "Blue Ridge Trail",
  area_name: "Blue Ridge Mountains",
  difficulty_rating: "moderate",
  length: 5.4,
  elevation_gain: 1200,
  route_type: "loop",
  trail_link: "https://www.alltrails.com/trail/blue-ridge-trail",
};

export const MOCK_HIKE_INFO = {
  id: 5,
  title: "Sunset Hike",
  date: "Wed, 12/01/2024",
  time: "6:00 PM",
  location: "Blue Ridge Parkway",
  comments: "Bring water and snacks.",
  status: "new",
  creator_id: 1,
  creator: "Annie McMahon",
  buttonMessage: "Join Hike"
};

export const MOCK_HIKE = {
  id: 5,
  title: "Sunset Hike",
  date: "2024-12-01",
  time: "18:00",
  location: "Blue Ridge Parkway",
  comments: "Bring water and snacks.",
  status: "new",
  creator_id: 1
};

export const MOCK_BIO_INFO = {
  user_name: "Hiker 123",
  bio: "Hello World",
};

export const MOCK_USER = {
  id: 2,
  email: "casualhiker@gmail.com",
  user_name: "Test User",
  avatar: "/newUser.png",
  bio: "Hello everyone!"
};



