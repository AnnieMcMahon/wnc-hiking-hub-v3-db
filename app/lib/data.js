import { sql } from '@vercel/postgres';

export async function fetchTrails() {
  try {
    const data = await sql`SELECT * FROM alltrails`;
    console.log(data.rows);
    return data.rows;
  } catch (error) {
    console.error('Database Error:', error);
    throw new Error('Failed to fetch trail data.');
  }
}
