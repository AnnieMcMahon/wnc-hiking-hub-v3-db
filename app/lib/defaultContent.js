export const defaultTrail = 
  {
    id: 1,
    trail_name: 'Triple Falls Trail',
    area_name: 'DuPont State Recreational Forest',
    difficulty_rating: 'moderate',
    length: 2.7,
    elevation_gain: 393,
    route_type: 'loop',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/triple-falls-trail'
  };

export const defaultUser = 
  {
    id: 1,
    email: "demo@gmail.com",
    password: "1234",
    user_name: "Demo User",
    avatar: "/newUser.png",
    bio: "Bio text goes here ... Lorem ipsum dolor sit amet consectetur adipisicing elit. Rem reprehenderit assumenda quibusdam iste eos exercitationem molestias nulla fuga repellat incidunt ullam cupiditate necessitatibus quidem debitis dolorum nesciunt labore quos, laboriosam tempora qui tempore quo velit! Libero quod eligendi at eum.",
    user_hikes: ["1", "2"]
  };

export const defaultHike = 
  {
    id: 1,
    creator_id: 1,
    trail_id: 1,
    title: "Sample Hike - Mid-week Morning Hike - Triple Falls",
    date: "2024-11-28",
    time: "10:00",
    location: "Triple Falls Parking Lot",
    comments: "Meet me by the blue CR-V. No dogs allowed. We will go slow and have time to take a lot of pictures.",
    status: "new"
  };