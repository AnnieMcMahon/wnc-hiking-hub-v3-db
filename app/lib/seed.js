const allTrails = [
  {
    id: 1,
    trail_name: 'Triple Falls Trail',
    area_name: 'DuPont State Recreational Forest',
    difficulty_rating: 'moderate',
    length: 2.7,
    elevation_gain: 393,
    route_type: 'loop',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/triple-falls-trail'
  },
  {
    id: 2,
    trail_name: 'Hard Times Loop',
    area_name: 'Pisgah National Forest',
    difficulty_rating: 'moderate',
    length: 6.4,
    elevation_gain: 531,
    route_type: 'loop',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/hardtimes-loop'
  },
  {
    id: 3,
    trail_name: 'Lake Powhatan via Bent Creek Trail',
    area_name: 'North Carolina Arboretum',
    difficulty_rating: 'easy',
    length: 5.5,
    elevation_gain: 196,
    route_type: 'out-and-back',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/lake-powhatan-loop-via-bent-creek-trail'
  },
  {
    id: 4,
    trail_name: 'Owl Ridge Trail',
    area_name: 'North Carolina Arboretum',
    difficulty_rating: 'moderate',
    length: 5.4,
    elevation_gain: 403,
    route_type: 'loop',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/owl-ridge-trail'
  },
  {
    id: 5,
    trail_name: 'Fletcher Loop',
    area_name: 'Bill Moore Community Park',
    difficulty_rating: 'easy',
    length: 3.1,
    elevation_gain: 22,
    route_type: 'loop',
    trail_link: 'https://www.alltrails.com/explore/trail/us/north-carolina/fletcher-park-loop'
  },
  {
    id: 6,
    trail_name: 'Looking Glass Rock Trail',
    area_name: 'Pisgah National Forest',
    difficulty_rating: 'moderate',
    length: 6.1,
    elevation_gain: 1683,
    route_type: 'out-and-back',
    trail_link: 'https://www.alltrails.com/trail/us/north-carolina/looking-glass-rock-trail'
  },
];

const sampleAppUsers = [
  {
    id: 1,
    email: "demo@gmail.com",
    password: "1234",
    user_name: "Demo User",
    avatar: "/newUser.png",
    bio: "Bio text goes here ... Lorem ipsum dolor sit amet consectetur adipisicing elit. Rem reprehenderit assumenda quibusdam iste eos exercitationem molestias nulla fuga repellat incidunt ullam cupiditate necessitatibus quidem debitis dolorum nesciunt labore quos, laboriosam tempora qui tempore quo velit! Libero quod eligendi at eum.",
    user_hikes: [1, 2, 3, 5, 6]
  },
  {
    id: 2,
    email: "anniemcmahon20@gmail.com",
    password: "1234",
    user_name: "Annie McMahon",
    avatar: "/avatar1.png",
    bio: "I've been hiking regularly for about five years now. I enjoy moderate hikes in a wooded area, with nice views or waterfalls, preferagly. My pace is about 2.5-3 mph. Join me on my next hike!",
    user_hikes: [2, 3, 4]
  },
];

const sampleHikes = [
  {
    id: 1,
    creator: 1,
    allTrailsId: 1,
    title: "Mid-week Morning Hike - Triple Falls",
    date: "2024-10-25",
    time: "10:00",
    location: "Triple Falls Parking Lot",
    comments: "Meet me by the blue CR-V. No dogs allowed. We will go slow and have time to take a lot of pictures."
  },
  {
    id: 2,
    creator: 2,
    allTrailsId: 2,
    title: "Afternoon Hike - Hard Times Loop",
    date: "2024-11-01",
    time: "14:00",
    location: "Hardtimes Rd, Asheville",
    comments: "Bring your dog! Limit 12 people. The trail is not as hard as the name indicates, don't worry! I'll be wearing a red jacket."
  },
  {
    id: 3,
    creator: 1,
    allTrailsId: 6,
    title: "Lake Powhatan via Bent Creek Trail",
    date: "2024-09-04",
    time: "09:00",
    location: "Arboretum hiking parking lot (left after the gate)",
    comments: "Be on time, we're leaving exactly at 9. We should be back before lunch, but bring a snack and plenty of water. We'll be taking a break midway."
  },
  {
    id: 4,
    creator: 2,
    allTrailsId: 4,
    title: "Owl Ridge Trail - laidback hike for everyone",
    date: "2026-09-21",
    time: "11:00",
    location: "NC Arboretum hiker's parking lot",
    comments: "This is a pretty easy hike. We can hang out at the arboretum's bistro afterwards."
  },
  {
    id: 5,
    creator: 1,
    allTrailsId: 5,
    title: "Fletcher Loop Fitness Hike",
    date: "2026-09-27",
    time: "15:00",
    location: "Bill Moore Community Park, near the dog park",
    comments: "This is a fitness hike. Only sign up if you can go 3 mph or faster. We will not wait for stragglers. You've been warned! Also, no dogs."
  },
  {
    id: 6,
    creator: 1,
    allTrailsId: 2,
    title: "Hard Times Loop",
    date: "2026-09-27",
    time: "10:00",
    location: "Hardtimes Rd",
    comments: "Bring  your lunch, plenty of water, and your dog, if you have one. See you on the trail!"
  },
];

export { allTrails, sampleAppUsers, sampleHikes };