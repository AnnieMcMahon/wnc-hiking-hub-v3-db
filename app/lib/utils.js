import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';

dayjs.extend(utc);
dayjs.extend(timezone);

const convertDate = (origDate) => {
  const date = dayjs(origDate).tz("America/New_York");
  const yyyy = date.year();
  const dddd = date.format("ddd");
  const mm = date.format("MM"); 
  const dd = date.format("DD");
  return `${dddd}, ${mm}/${dd}/${yyyy}`;
};

const convertTime = (time) => {
  let hh = parseInt(time.slice(0, 2));
  let mm = time.slice(3, 5);
  let ampm = hh >= 12 ? 'PM' : 'AM';
  hh = hh % 12;
  hh = hh ? hh : 12;
  const formattedTime = hh + ":" + mm + " " + ampm;
  return formattedTime;
};

export { convertDate, convertTime };




