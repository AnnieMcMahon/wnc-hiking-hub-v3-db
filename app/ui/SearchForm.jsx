import { ANY_AREA, ANY_DIFFICULTY, ANY_LENGTH } from "@/app/lib/constants";

export default function SearchForm({
  searchByArea,
  searchByDifficulty,
  searchByLength,
}) {
  return (
    <form>
      <label htmlFor="area_name">Area Name: </label>
      <select name="area_name" id="area_name" onChange={searchByArea}>
        <option value={ANY_AREA}>{ANY_AREA}</option>
        <option value="DuPont State Recreational Forest">
          DuPont State Recreational Forest
        </option>
        <option value="Pisgah National Forest">Pisgah National Forest</option>
        <option value="North Carolina Arboretum">
          North Carolina Arboretum
        </option>
        <option value="Nantahala Forest">Nantahala Forest</option>
      </select>
      <br />
      <label htmlFor="difficulty">Difficulty: </label>
      <select name="difficulty" id="difficulty" onChange={searchByDifficulty}>
        <option value={ANY_DIFFICULTY}>{ANY_DIFFICULTY}</option>
        <option value="Easy">Easy</option>
        <option value="Moderate">Moderate</option>
        <option value="Strenuous">Strenuous</option>
      </select>
      <label htmlFor="length"> Length: </label>
      <select name="length" id="length" onChange={searchByLength}>
        <option value={ANY_LENGTH}>{ANY_LENGTH}</option>
        <option value="Short">Shorter than 3 miles</option>
        <option value="Medium">From 3 to 6 miles</option>
        <option value="Long">Longer than 6 miles</option>
      </select>
      <br />
    </form>
  );
}
