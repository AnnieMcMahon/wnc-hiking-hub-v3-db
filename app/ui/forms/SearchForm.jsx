import { AREAS, DIFFICULTIES, LENGTHS, ANY_AREA, ANY_DIFFICULTY, ANY_LENGTH } from "@/app/lib/constants";

export default function SearchForm({ onSearch = () => {} }) {
  const handleSearch = (e) => {
    const { name, value } = e.target;
    onSearch(name, value);
  };
  return (
    <form className="search-form text-sm">
      <label htmlFor="area">Area Name: </label>
      <select
        name="area"
        id="area"
        onChange={handleSearch}
      >
        <option key={ANY_AREA} value={ANY_AREA}>{ANY_AREA}</option>
        {AREAS.map((area) => (
          <option key={area} value={area}>{area}</option>
        ))}
      </select>
      <br />
      <label htmlFor="difficulty">Difficulty Rating: </label>
      <select
        name="difficulty"
        id="difficulty"
        onChange={handleSearch}
      >
        <option key={ANY_DIFFICULTY} value={ANY_DIFFICULTY}>
          {ANY_DIFFICULTY}</option>
        {DIFFICULTIES.map((diff) => (
          <option key={diff} value={diff}>{diff}</option>
        ))}
      </select>
      <label htmlFor="length"> Length: </label>
      <select
        name="length"
        id="length"
        onChange={handleSearch}
      >
        <option key={ANY_LENGTH} value={ANY_LENGTH}>{ANY_LENGTH}</option>
        {LENGTHS.map((length) => (
          <option key={length} value={length}>{length}</option>
        ))}
      </select>
      <br />
    </form>
  );
}
