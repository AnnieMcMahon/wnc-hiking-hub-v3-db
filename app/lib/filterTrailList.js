import { fetchAllTrails } from "@/app/api/data";

export async function filterTrailList(area, difficulty, length) {
  let newList = await fetchAllTrails();
  if (area !== "Anywhere in WNC") {
    newList = newList.filter(trail => trail.area == area)
  }
  if (difficulty !== "Any") {
    newList = newList.filter(trail => trail.difficulty == difficulty)
  }
  if (length !== "Any length") {
    newList = newList.filter(trail => 
      (length == "Short" && Number(trail.length) < 3) ||
      (length == "Long" && Number(trail.length > 6)) ||
        (length == "Medium" && Number(trail.length) >=3 && Number(trail.length) <= 6))
  }
  return newList;
};