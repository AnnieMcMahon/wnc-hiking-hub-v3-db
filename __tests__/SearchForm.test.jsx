import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SearchForm from "@/app/ui/forms/SearchForm";
import { ANY_AREA, ANY_DIFFICULTY, ANY_LENGTH } from "@/app/lib/constants";

describe("SearchForm", () => {
  describe("rendering", () => {
    it("renders a SearchForm component", () => {
      render(<SearchForm />);
      expect(screen.getByText(/area name/i)).toBeInTheDocument();
    });

    it("renders default data correctly", () => {
      render(<SearchForm />);
      expect(screen.getByText("Anywhere in WNC")).toBeInTheDocument();
      expect(screen.getByText("Any difficulty")).toBeInTheDocument();
      expect(screen.getByText("Any length")).toBeInTheDocument();
    });

    it("renders dropdowns with default values", () => {
      render(<SearchForm />);
      const areaDropdown = screen.getByRole("combobox", { name: /area name/i });
      const difficultyDropdown = screen.getByRole("combobox", {
        name: /difficulty rating/i,
      });
      const lengthDropdown = screen.getByRole("combobox", { name: /length/i });
      expect(areaDropdown).toHaveValue(ANY_AREA);
      expect(difficultyDropdown).toHaveValue(ANY_DIFFICULTY);
      expect(lengthDropdown).toHaveValue(ANY_LENGTH);
    });

    it("renders all area options correctly", () => {
      render(<SearchForm />);
      const areaDropdown = screen.getByRole("combobox", { name: /area name/i });
      const options = Array.from(areaDropdown.options).map(
        (option) => option.text
      );
      expect(options).toEqual([
        "Anywhere in WNC",
        "DuPont State Recreational Forest",
        "Pisgah National Forest",
        "North Carolina Arboretum",
        "Nantahala Forest",
      ]);
    });
  });

  describe("functional", () => {
    it("calls onSearch with the correct arguments when a dropdown value changes", async () => {
      const mockOnSearch = jest.fn();
      render(<SearchForm onSearch={mockOnSearch} />);
      const user = userEvent.setup();
      const areaDropdown = screen.getByRole("combobox", { name: /area name/i });
      await user.selectOptions(areaDropdown, "Pisgah National Forest");
      expect(mockOnSearch).toHaveBeenCalledWith(
        "area",
        "Pisgah National Forest"
      );
      const difficultyDropdown = screen.getByRole("combobox", {
        name: /difficulty rating/i,
      });
      await user.selectOptions(difficultyDropdown, "moderate");
      expect(mockOnSearch).toHaveBeenCalledWith("difficulty", "moderate");
      const lengthDropdown = screen.getByRole("combobox", { name: /length/i });
      await user.selectOptions(lengthDropdown, "From 3 to 6 miles");
      expect(mockOnSearch).toHaveBeenCalledWith("length", "From 3 to 6 miles");
    });
  });
});
