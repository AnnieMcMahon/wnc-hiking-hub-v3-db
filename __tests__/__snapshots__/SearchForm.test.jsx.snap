// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm matches the snapshot 1`] = `
<div>
  <form
    class="search-form"
  >
    <label
      for="area"
    >
      Area Name: 
    </label>
    <select
      id="area"
      name="area"
    >
      <option
        value="Anywhere in WNC"
      >
        Anywhere in WNC
      </option>
      <option
        value="DuPont State Recreational Forest"
      >
        DuPont State Recreational Forest
      </option>
      <option
        value="Pisgah National Forest"
      >
        Pisgah National Forest
      </option>
      <option
        value="North Carolina Arboretum"
      >
        North Carolina Arboretum
      </option>
      <option
        value="Nantahala Forest"
      >
        Nantahala Forest
      </option>
    </select>
    <br />
    <label
      for="difficulty"
    >
      Difficulty Rating: 
    </label>
    <select
      id="difficulty"
      name="difficulty"
    >
      <option
        value="Any difficulty"
      >
        Any difficulty
      </option>
      <option
        value="easy"
      >
        easy
      </option>
      <option
        value="moderate"
      >
        moderate
      </option>
      <option
        value="hard"
      >
        hard
      </option>
    </select>
    <label
      for="length"
    >
       Length: 
    </label>
    <select
      id="length"
      name="length"
    >
      <option
        value="Any length"
      >
        Any length
      </option>
      <option
        value="Shorter than 3 miles"
      >
        Shorter than 3 miles
      </option>
      <option
        value="From 3 to 6 miles"
      >
        From 3 to 6 miles
      </option>
      <option
        value="Longer than 6 miles"
      >
        Longer than 6 miles
      </option>
    </select>
    <br />
  </form>
</div>
`;
